File: sasu-mobile/lib/data/models/insurance.dart
/// Insurance Model
///
/// Represents an insurance policy and its future benefit
/// Used to calculate family protection and future maturity values

class Insurance {
  final int? id;
  final String policyName;
  final String type;
  final String provider;
  final double coverageAmount;
  final double? premiumAmount;
  final String? premiumFrequency;
  final String? startDate;
  final int? maturityYear;
  final double? maturityBenefit;
  final String beneficiary;
  final String? description;
  final bool active;

  Insurance({
    this.id,
    required this.policyName,
    required this.type,
    required this.provider,
    required this.coverageAmount,
    this.premiumAmount,
    this.premiumFrequency,
    this.startDate,
    this.maturityYear,
    this.maturityBenefit,
    required this.beneficiary,
    this.description,
    this.active = true,
  });

  factory Insurance.fromJson(Map<String, dynamic> json) {
    return Insurance(
      id: json['id'],
      policyName: json['policyName'] ?? '',
      type: json['type'] ?? '',
      provider: json['provider'] ?? '',
      coverageAmount: (json['coverageAmount'] ?? 0).toDouble(),
      premiumAmount: json['premiumAmount']?.toDouble(),
      premiumFrequency: json['premiumFrequency'],
      startDate: json['startDate'],
      maturityYear: json['maturityYear'],
      maturityBenefit: json['maturityBenefit']?.toDouble(),
      beneficiary: json['beneficiary'] ?? '',
      description: json['description'],
      active: json['active'] ?? true,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'policyName': policyName,
      'type': type,
      'provider': provider,
      'coverageAmount': coverageAmount,
      'premiumAmount': premiumAmount,
      'premiumFrequency': premiumFrequency,
      'startDate': startDate,
      'maturityYear': maturityYear,
      'maturityBenefit': maturityBenefit,
      'beneficiary': beneficiary,
      'description': description,
      'active': active,
    };
  }
}

File: sasu-backend/src/main/java/com/sasu/family/model/Insurance.java
package com.sasu.family.model;

import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.math.BigDecimal;
import java.time.LocalDate;

/**
 * Insurance Model
 *
 * Represents an insurance policy that provides protection or future benefit.
 *
 * Examples:
 * - Life insurance
 * - Medical insurance
 * - Education plan
 * - Vehicle insurance
 *
 * Key focus:
 * - Coverage amount
 * - Maturity year
 * - Beneficiaries
 *
 * Used for:
 * - Family protection score
 * - Future benefit projections
 */
@Entity
@Table(name = "insurance")
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Insurance {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false)
    private String policyName;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private InsuranceType type;

    @Column(nullable = false)
    private String provider;

    @Column(nullable = false, precision = 15, scale = 2)
    private BigDecimal coverageAmount;

    @Column(precision = 15, scale = 2)
    private BigDecimal premiumAmount;

    @Enumerated(EnumType.STRING)
    private PremiumFrequency premiumFrequency;

    private LocalDate startDate;

    private Integer maturityYear;

    @Column(precision = 15, scale = 2)
    private BigDecimal maturityBenefit;

    @Column(nullable = false)
    private String beneficiary;

    @Column(length = 1000)
    private String description;

    @Column(nullable = false)
    @Builder.Default
    private Boolean active = true;

    public enum InsuranceType {
        LIFE,
        MEDICAL,
        EDUCATION,
        VEHICLE,
        HOME,
        OTHER
    }

    public enum PremiumFrequency {
        MONTHLY,
        QUARTERLY,
        HALF_YEARLY,
        YEARLY
    }
}

